src/security/KeyData.cc

```diff
@@ -96,14 +94,37 @@ Security::KeyData::loadX509ChainFromFile()
         return;
     }
 
-    if (X509_check_issued(cert.get(), cert.get()) == X509_V_OK)
-        debugs(83, 5, "Certificate is self-signed, will not be chained");
-    else {
+    if (X509_check_issued(cert.get(), cert.get()) == X509_V_OK) {
+        char *nameStr = X509_NAME_oneline(X509_get_subject_name(cert.get()), nullptr, 0);
+        debugs(83, DBG_PARSE_NOTE(2), "Certificate is self-signed, will not be chained: " << nameStr);
+        OPENSSL_free(nameStr);
+    } else {
+        debugs(83, DBG_PARSE_NOTE(3), "Using certificate chain in " << certFile);
         // and add to the chain any other certificate exist in the file
-        while (X509 *ca = PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr)) {
-            // XXX: self-signed check should be applied to all certs loaded.
-            // XXX: missing checks that the chained certs are actually part of a chain for validating cert.
-            chain.emplace_front(Security::CertPointer(ca));
+        CertPointer latestCert = cert;
+
+        while (auto ca = PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr)) {
+            // get Issuer name of the cert for debug display
+            char *nameStr = X509_NAME_oneline(X509_get_subject_name(ca), nullptr, 0);
+
+            // self-signed certificates are not valid in a sent chain
+            if (X509_check_issued(ca, ca) == X509_V_OK) {
+                debugs(83, DBG_PARSE_NOTE(2), "CA " << nameStr << " is self-signed, will not be chained: " << nameStr);
```

Comment:
The RFC 5246 says about certificate chain:

> Each following certificate MUST directly certify the one preceding it.  Because certificate validation requires that root keys be distributed independently, **the self-signed certificate that specifies the root certificate authority MAY be omitted from the chain**, under the assumption that the remote end must already possess it in order to validate it in any case.

My sense is that we should allow the CA root certificates. Just it should be the last one.

Reply:
If there are no known use cases where a single root certificate in the right position breaks things, then I agree that we should allow it. Otherwise, I am not sure what the correct solution is.

Reply:
I'm not sure what the exact tool output was earlier in my experimentation to find the actual bug. It seems to be working now for the tests I am doing regularly. So hiding the self-signed checks inside pre-processor blocks in hopes that we can remove them later instead of having anyone need to enable.